### PR TITLE
Apply assorted Repo-Review suggestions

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
Mostly stricter pytest and MyPy settings.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: zarr/storage.py [hunk 1]
```diff
@@ -446,7 +446,11 @@ def init_array(
 
     if store_version == 3 and "zarr.json" not in store:
         # initialize with default zarr.json entry level metadata
-        store["zarr.json"] = store._metadata_class.encode_hierarchy_metadata(None)  # type: ignore
+        # fmt: off
```
Comment (DimitriPapadopoulos) [inline zarr/storage.py:449]: That's the only way I have found to ensure `black` doesn't create lines > 100 characters, while not moving `# type: ignore[union-attr]` to an unrelated line (the last line with the parenthesis).

Suggestions welcome.

Comment (d-v-b) [reply zarr/storage.py:449]: the ongoing v3 effort will be stripping out a lot of the old v3 logic, so do whatever appeases the linter here, no matter how ugly it is

File: zarr/storage.py [hunk 2]
```diff
@@ -2497,7 +2505,7 @@ def _cache_value(self, key: Path, value):
         value_size = buffer_size(value)
         # check size of the value against max size, as if the value itself exceeds max
         # size then we are never going to cache it
-        if self._max_size is None or value_size <= self._max_size:
+        if self._max_size is None or value_size <= self._max_size:  # type: ignore[redundant-expr]
```
Comment (DimitriPapadopoulos) [inline zarr/storage.py:2508]: `self._max_size` is initialised here:
https://github.com/zarr-developers/zarr-python/blob/f58065b221452acd70235902ad59d920da6fb02f/zarr/storage.py#L2406-L2408
and here:
https://github.com/zarr-developers/zarr-python/blob/f58065b221452acd70235902ad59d920da6fb02f/zarr/storage.py#L2417-L2420

In the first case it shouldn't be `None`, but then typing cannot prevent giving the `max_size` parameter the `None` value. In the second case, no typing is involved.?Not sure why ruff thinks `self._max_size is None` is redundant.

**Comment by jhamman** [comment]
@DimitriPapadopoulos - thanks for moving your focus over to the `v3` branch. I'm going to close this in favor of the contributions you've already made there. 🙏
